crates/router/src/core/payments/operations/payment_session.rs

```diff
@@ -310,131 +308,94 @@ where
+                        api_models::enums::PaymentMethodType::ApplePay
+                        | api_models::enums::PaymentMethodType::GooglePay => {
+                            Ok(connector_tokeninfo.0.as_str())
                         }
-                    })
-            })
-            .map(|filtered_connector| filtered_connector.connector_name.clone())
-            .collect::<HashSet<String>>();
-
-        let given_wallets = request.wallets.clone();
-
-        let connectors_data = if !given_wallets.is_empty() {
-            // Create connectors for provided wallets
-            let mut connectors_data = Vec::with_capacity(supported_connectors.len());
-            for wallet in given_wallets {
-                let (connector_name, connector_type) = match wallet {
-                    api_enums::SupportedWallets::Gpay => ("adyen", api::GetToken::Metadata),
-                    api_enums::SupportedWallets::ApplePay => ("applepay", api::GetToken::Connector),
-                    api_enums::SupportedWallets::Paypal => ("braintree", api::GetToken::Connector),
-                    api_enums::SupportedWallets::Klarna => ("klarna", api::GetToken::Connector),
-                };
-
-                // Check if merchant has enabled the required merchant connector account
-                if session_token_from_metadata_connectors.contains(connector_name)
-                    || normal_connector_names.contains(connector_name)
-                {
-                    connectors_data.push(api::ConnectorData::get_connector_by_name(
-                        connectors,
-                        connector_name,
-                        connector_type,
-                    )?);
+                        _ => Err(report!(errors::ApiErrorResponse::InternalServerError))
+                            .attach_printable("Payment Method not supported for session token"),
+                    }?;
+                    if connector_tokeninfo.1 == api::GetToken::from(payment_method_type) {
+                        let connector_details = api::ConnectorData::get_connector_by_name(
+                            connectors,
+                            connector_name,
+                            connector_tokeninfo.1.to_owned(),
+                        )?;
+                        connectors_data.push(connector_details);
+                    }
                 }
             }
             connectors_data
         } else {
-            // Create connectors for all enabled wallets
-            let mut connectors_data = Vec::with_capacity(
-                normal_connector_names.len() + session_token_from_metadata_connectors.len(),
-            );
+            let mut connectors_data = Vec::new();
 
-            for connector_name in normal_connector_names {
-                let connector_data = api::ConnectorData::get_connector_by_name(
+            for connector_tokeninfo in &connector_and_supporting_wallet {
```

Comment:
since this vector is not being used again, consume it in this for loop instead of taking the reference